tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java

```diff
@@ -138,14 +219,205 @@ public GraphOperation operation(String name) {
     return new OperationIterator(this);
   }
 
+  private GraphOperation graphOp(Operand<?> operand) {
+    checkInput(operand);
+    return (GraphOperation) operand.op();
+  }
+
+  /**
+   * Finds the operations used to produce {@code outputs} from {@code inputs}, or throws if that is not possible.
+   * Includes control dependencies.
+   *
+   * @param inputs the inputs of the subgraph.  Must be from single output ops.  May not be null.
+   * @param outputs the outputs of the subgraph.  May not be null.
+   * @param allowNoInputBodyOps whether to allow 0-input ops in the body.  For more specificy use {@link
+   * #completeSubgraph(Set, Set, Set, Set)}.
+   * @return the set of operations needed to calculate outputs from inputs, including outputs and inputs
+   * @throws IllegalStateException if outputs depends on ops outside of the subgraph (i.e. is not calculable based
+   * solely on inputs)
+   * @see #completeSubgraph(Set, Set, Set, Set)
+   */
+  public Set<GraphOperation> completeSubgraph(Set<Operand<?>> inputs, Set<Operand<?>> outputs,
+      boolean allowNoInputBodyOps) {
+    return completeSubgraph(inputs, outputs, null, allowNoInputBodyOps ? Collections.emptySet() : null);
+  }
+
+  /**
+   * Finds the operations used to produce {@code outputs} from {@code inputs}, or throws if that is not possible.
+   * Includes control dependencies.
+   *
+   * If both {@code allowedNoInputBodyOps} and {@code forbiddenNoInputBodyOps} are {@code null}, forbids 0-input ops in
+   * the body. To allow all ops in the body, use {@code null} for {@code allowedNoInputBodyOps} and the empty set for
+   * {@code forbiddenNoInputBodyOps}.
+   *
+   * @param inputs the inputs of the subgraph.  Must be from single output ops.  May not be null.
+   * @param outputs the outputs of the subgraph.  May not be null.
+   * @param allowedNoInputBodyOps types of ops to allow as 0-input ops in the body.  Allows all (except {@code
+   * forbiddenNoInputBodyOps}) if null.
+   * @param forbiddenNoInputBodyOps types of ops to never allow as 0-input ops in the body.  Forbids all (except {@code
+   * allowedNoInputBodyOps}) if null.
+   * @return the set of operations needed to calculate outputs from inputs, including outputs and inputs
+   * @throws IllegalStateException if outputs depends on ops outside of the subgraph (i.e. is not calculable based
+   * solely on inputs)
+   * @see #completeSubgraph(Set, Set, boolean)
+   */
+  public synchronized Set<GraphOperation> completeSubgraph(Set<Operand<?>> inputs, Set<Operand<?>> outputs,
+      Set<String> allowedNoInputBodyOps, Set<String> forbiddenNoInputBodyOps) {
```

Comment:
These two last extra parameters obfuscate a bit the overall solution, which would be very simple without them. Why can't we simply include all ops we encounter without this logic, whether they have inputs available or not, and let the caller filter them out after if he needs to?

Reply:
Yeah, that's fine.  I don't like how easy it is to end up including things upstream from the inputs in the body then, but I don't see much of a way around it, since we can't just forbid it altogether.